### ISSUE TITLE
ERROR: MethodError: no method matching typeof(exp)()

### ISSUE THREAD

**Issue by jbrea** [ThummeTo/ForwardDiffChainRules.jl#4]
Cool and useful package! Looking forward to its registration.
While testing I ran into the following issue.
Here is a somewhat contrived MWE:
```julia
using ChainRulesCore, ForwardDiff, ForwardDiffChainRules, ChainRulesTestUtils
g(f, x; kwargs...) = f(x; kwargs...)
function ChainRulesCore.frule((_, _, Δx), ::typeof(g), f, x; kwargs...)
    y = g(f, x; kwargs...)
    return y, f(x) * Δx
end
test_frule(g, exp, .4)
@ForwardDiff_frule g(f, x::ForwardDiff.Dual)
julia> ForwardDiff.derivative(x -> g(exp, x), .3)
ERROR: MethodError: no method matching typeof(exp)()
Stacktrace:
  [1] structfromnt(#unused#::Type{typeof(exp)}, x::NamedTuple{(), Tuple{}})
    @ NamedTupleTools ~/.julia/packages/NamedTupleTools/2Wwof/src/NamedTupleTools.jl:118
  [2] (::DifferentiableFlatten.var"#21#22"{typeof(exp), DifferentiableFlatten.var"#unflatten_to_NamedTuple#13"{NamedTuple{(), Tuple{}}, DifferentiableFlatten.var"#61#62"}})(y::Vector{Float64})
    @ DifferentiableFlatten ~/.julia/packages/DifferentiableFlatten/C0Hvn/src/DifferentiableFlatten.jl:111
  [3] (::DifferentiableFlatten.Unflatten{typeof(exp), DifferentiableFlatten.var"#21#22"{typeof(exp), DifferentiableFlatten.var"#unflatten_to_NamedTuple#13"{NamedTuple{(), Tuple{}}, DifferentiableFlatten.var"#61#62"}}})(x::Vector{Float64})
    @ DifferentiableFlatten ~/.julia/packages/DifferentiableFlatten/C0Hvn/src/DifferentiableFlatten.jl:218
  [4] (::DifferentiableFlatten.var"#9#12"{Vector{Float64}})(x_back::DifferentiableFlatten.Unflatten{typeof(exp), DifferentiableFlatten.var"#21#22"{typeof(exp), DifferentiableFlatten.var"#unflatten_to_NamedTuple#13"{NamedTuple{(), Tuple{}}, DifferentiableFlatten.var"#61#62"}}}, l::Int64, s::Int64)
    @ DifferentiableFlatten ~/.julia/packages/DifferentiableFlatten/C0Hvn/src/DifferentiableFlatten.jl:76
  [5] map
    @ ./tuple.jl:266 [inlined]
  [6] (::DifferentiableFlatten.var"#unflatten_to_Tuple#11"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{DifferentiableFlatten.Unflatten{typeof(exp), DifferentiableFlatten.var"#21#22"{typeof(exp), DifferentiableFlatten.var"#unflatten_to_NamedTuple#13"{NamedTuple{(), Tuple{}}, DifferentiableFlatten.var"#61#62"}}}, DifferentiableFlatten.var"#unflatten_to_Real#1"}})(v::Vector{Float64})
    @ DifferentiableFlatten ~/.julia/packages/DifferentiableFlatten/C0Hvn/src/DifferentiableFlatten.jl:75
  [7] g(f::Function, x::ForwardDiff.Dual{ForwardDiff.Tag{var"#32#33", Float64}, Float64, 1})
    @ Main ~/.julia/packages/ForwardDiffChainRules/Hgko1/src/ForwardDiffChainRules.jl:54
  [8] (::var"#32#33")(x::ForwardDiff.Dual{ForwardDiff.Tag{var"#32#33", Float64}, Float64, 1})
    @ Main ./REPL[19]:1
  [9] derivative(f::var"#32#33", x::Float64)
    @ ForwardDiff ~/.julia/packages/ForwardDiff/pDtsf/src/derivative.jl:14
 [10] top-level scope
    @ REPL[19]:1
```

**Comment by mohdibntarek** [comment]
Should be fixed with DifferentiableFlatten 0.1.1. I will make a PR to bump its lower bound here once it's registered https://github.com/JuliaRegistries/General/pull/70696.

**Comment by mohdibntarek** [comment]
https://github.com/ThummeTo/ForwardDiffChainRules.jl/pull/6

**Comment by ThummeTo** [comment]
This is fixed in v0.1.1:

```
test_frule(g, exp, .4)
Test Summary:                        | Pass  Total  Time
test_frule: g on typeof(exp),Float64 |    4      4  1.6s
Test.DefaultTestSet("test_frule: g on typeof(exp),Float64", Any[], 4, false, false, true, 1.666364057337e9, 1.666364058946e9)
```

(please re-open if there are still issues)
